File: 0124_maxPathSum.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def maxPathSum(self, root: Optional[TreeNode]) -> int:
        self.maxSum = -float("inf")
        self.dfs(root)
        return self.maxSum

    def dfs(self, node):
        if not node:
            return 0
        left = self.dfs(node.left)
        right = self.dfs(node.right)
        self.maxSum = max(self.maxSum, left + right + node.val)
        res = max(max(left, right) + node.val, node.val)
        return res if res > 0 else 0

File: 0001-0500/0124_maxPathSum.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public int max = Integer.MIN_VALUE;

    public int maxPathSum(TreeNode root) {
        dfs(root);
        return max;
    }

    public int dfs(TreeNode root) {
        if (root.left == null && root.right == null) {
            max = Math.max(max, root.val);
            return Math.max(root.val, 0);
        }

        int left = (root.left != null) ? dfs(root.left) : 0;
        int right = (root.right != null) ? dfs(root.right) : 0;
        max = Math.max(max, left + right + root.val);
        return Math.max(Math.max(left, right) + root.val, 0);
    }
}
